solrmonitor/zkdispatcher.go

```diff
@@ -24,24 +24,49 @@ import (
 	"github.com/samuel/go-zookeeper/zk"
 )
 
-// Handle a zk event.  Optionally return a new event channel to automatically re-register.
+// ugh... seems silly to have to define this
+const maxInt = int(^uint(0) >> 1)
+
+// A function that handles a zk event. Optionally returns a new event channel to automatically
+// re-register.
 type ZkEventHandler func(zk.Event) <-chan zk.Event
 
-// Monitors many zk.Event channels on a single goroutine.
+// An interface that handles a zk event. Optionally returns a new event channel to automatically
+// re-register.
+type ZkEventCallback interface {
+	Handle(zk.Event) <-chan zk.Event
+}
+
+// Adapts a ZkEventHandler function to the ZkEventCallback interface.
+type zkEventHandlerAdapter struct {
+	handler ZkEventHandler
+}
+
+func (a *zkEventHandlerAdapter) Handle(event zk.Event) <-chan zk.Event {
+	return a.handler(event)
+}
+
+// Monitors many zk.Event channels. Dispatches handling to other goroutines allowing handlers to
+// run in parallel. A single ZkEventCallback, however, will always be invoked as if from the
+// same goroutine. So a single callback can be used to watch multiple channels and it will receive
+// events in a way that does not require it to synchronize with other possible invocations.
```

Comment:
Q: how much complexity could be removed by relaxing this?  I would argue that most callbacks will need internal synchronization *anyway* just to sync up with external callers.